cmd/ptpcheck/cmd/sources.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright (c) Facebook, Inc. and its affiliates.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/facebook/time/cmd/ptpcheck/checker"
+	"github.com/facebook/time/phc"
+	ptp "github.com/facebook/time/ptp/protocol"
+
+	"github.com/olekukonko/tablewriter"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var (
+	sourcesNoDNSFlag bool
+)
+
+func init() {
+	RootCmd.AddCommand(sourcesCmd)
+	sourcesCmd.Flags().StringVarP(&rootServerFlag, "server", "S", "/var/run/ptp4l", "server to connect to")
+	sourcesCmd.Flags().BoolVarP(&sourcesNoDNSFlag, "no-resolving", "n", false, "disable resolving of IP addresses to hostnames")
+}
+
+func sourcesRun(server string, noDNS bool) error {
+	c, cleanup, err := checker.PrepareClient(server)
+	defer cleanup()
+	if err != nil {
+		return fmt.Errorf("preparing connection: %w", err)
+	}
+	ppn, err := c.PortPropertiesNP()
+	if err != nil {
+		return fmt.Errorf("getting PORT_PROPERTIES_NP from ptp4l: %w", err)
+	}
+	cds, err := c.CurrentDataSet()
+	if err != nil {
+		return fmt.Errorf("getting CURRENT_DATA_SET from ptp4l: %w", err)
+	}
+	tsn, err := c.TimeStatusNP()
+	if err != nil {
+		return fmt.Errorf("getting TIME_STATUS_NP from ptp4l: %w", err)
+	}
+	tlv, err := c.UnicastMasterTableNP()
+	if err != nil {
+		return fmt.Errorf("getting UNICAST_MASTER_TABLE_NP from ptp4l: %w", err)
+	}
+	// obtain time from clock that ptp4l uses
+	var currentTime time.Time
+	if ppn.Timestamping == ptp.TimestampingHardware {
+		currentTime, err = phc.Time(string(ppn.Interface), phc.MethodIoctlSysOffsetExtended)
+		if err != nil {
+			log.Errorf("No PHC time data available: %v", err)
+		}
+	} else {
+		currentTime = time.Now()
+	}
```

Comment:
If I understand this correctly in 1 case you will get TAI and in another UTC. Will 37 seconds offset be a problem?

Reply:
No, ptp4l reports time of last sync as seen by the clock that is used for timestamping. So if HW timestamping is used - it returns time from PHC (in TAI), and we get time from it as well. And if it uses software timestamping, it takes time from sys clock, and so do we.